Invalidate variations upon generation and fix formatted name

louwie17 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Fixes two issues:
- The variation name not showing correctly when more then two attributes are used
- The variation names not being updated after navigating back and generating new variations.

Closes # .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Make sure feature `New product editor` is enabled under `/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`
2. Go to `/wp-admin/admin.php?page=wc-admin&path=/add-product&tab=variations` and add some attributes (at-least more then 1 )
3. Add at-least 3 or more variation options
4. Wait until variations get generated
5. Then publish the product and save its `productId` and one `variationId` you like
6. Then click **Edit** on the first variation in the list
7. Note that the variation name and name in the switcher contain the correct list of attributes ( and not the product name )
8. Click **Main product**
9. Remove or add one a variation option and wait for it to generate them again.
10. Click edit again
11. The variation name should be correct with the added or removed attribute.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

louwie17 committed:
Invalidate variations upon generation and fix formatted name

mdperez86 commented on packages/js/product-editor/src/hooks/use-product-variations-helper.ts:
```diff
@@ -111,6 +113,15 @@ export function useProductVariationsHelper() {
 					invalidateResolutionForStoreSelector(
 						'getProductVariations'
 					);
+					if ( variations && variations.length > 0 ) {
```
Well this does not work when there are old variations in cache, when invalidating the old variations the a request is made and the server responds with 404. Also we might need to invalidate the parent product request here too, since generate variations affect the `variations` property of the product which is used in the variation switcher.

louwie17 replied:
Hmm yeah I could do the `invalidateResolutionForStoreSelector( 'getEntityRecord' )` which would do it for all of them, but I was trying to avoid that.

mdperez86 replied:
I know but if we don't do it then the variation ids from the parent product never get updated, so when we land to the single variation page, that list is being used for the swiper and it'll be outdated. Once way to avoid that is updating the local copy of the variation ids within the cached parent product right after we generate variations, which is going to force the user to save the product before leaving the page.

louwie17 replied:
I included the invalidation of the product here.

mdperez86 reviewed (changes requested):
Nice work @louwie17. I left one comment.

louwie17 committed:
Add aditional invalidates to keep cache up to date

louwie17 committed:
Update formatted name to always return attribute term name list

louwie17 commented:
@mdperez86 this is good for a re-review, I added additional cache validation in this commit: https://github.com/woocommerce/woocommerce/pull/40905/commits/1fa9047b84c7e9be24160eb366bdaab6cc992d9c one's specifically for when we delete variations to invalidate the product and the variation it-self.
I also found a small bug with the variation name, which would be blank if it just included two attributes, which I fixed in this commit: https://github.com/woocommerce/woocommerce/pull/40905/commits/ce363087db702b344bfda88599703411a3c21746

louwie17 commented:
I do wonder if there is some better way we can determine dependencies between actions and stores and not have to be as conscious about invalidating for some of these calls/actions, where the store it-self handles it for us.

mdperez86 reviewed (approved):
God job @louwie17 LGTM!!!
